java/ql/lib/semmle/code/java/security/InsufficientKeySize.qll

```diff
@@ -0,0 +1,191 @@
+private int getMinAsymEcKeySize() { result = 256 }
+
+/** Returns the minimum recommended key size for AES algorithms. */
+private int getMinSymKeySize() { result = 128 }
+
+/** Returns the integer value of a given DataFlow::Node. */
+private int getNodeIntValue(DataFlow::Node node) {
+  result = node.asExpr().(IntegerLiteral).getIntValue()
+}
+
+/** Returns the key size from an EC algorithm's curve name string */
+bindingset[algorithm]
+private int getEcKeySize(string algorithm) {
+  algorithm.matches("sec%") and // specification such as "secp256r1"
+  result = algorithm.regexpCapture("sec[p|t](\\d+)[a-zA-Z].*", 1).toInt()
+  or
+  algorithm.matches("X9.62%") and //specification such as "X9.62 prime192v2"
+  result = algorithm.regexpCapture("X9\\.62 .*[a-zA-Z](\\d+)[a-zA-Z].*", 1).toInt()
+  or
+  (algorithm.matches("prime%") or algorithm.matches("c2tnb%")) and //specification such as "prime192v2"
+  result = algorithm.regexpCapture(".*[a-zA-Z](\\d+)[a-zA-Z].*", 1).toInt()
+}
+
+// ************************************ SINKS ************************************
+/** A sink for an insufficient key size used in RSA, DSA, and DH algorithms. */
+private class AsymmetricNonEcSink extends InsufficientKeySizeSink {
+  AsymmetricNonEcSink() {
+    exists(AsymmetricInitMethodAccess ma, AsymmetricKeyGenerator kg |
+      kg.getAlgoName().matches(["RSA", "DSA", "DH"]) and
+      DataFlow::localExprFlow(kg, ma.getQualifier()) and
+      this.asExpr() = ma.getKeySizeArg()
+    )
+    or
+    exists(AsymmetricNonEcSpec spec | this.asExpr() = spec.getKeySizeArg())
+  }
+
+  override predicate hasState(DataFlow::FlowState state) {
+    state = getMinAsymNonEcKeySize().toString()
+  }
+}
+
+/** A sink for an insufficient key size used in elliptic curve (EC) algorithms. */
+private class AsymmetricEcSink extends InsufficientKeySizeSink {
+  AsymmetricEcSink() {
+    exists(AsymmetricInitMethodAccess ma, AsymmetricKeyGenerator kg |
+      kg.getAlgoName().matches("EC%") and
+      DataFlow::localExprFlow(kg, ma.getQualifier()) and
+      this.asExpr() = ma.getKeySizeArg()
+    )
+    or
+    exists(AsymmetricEcSpec s | this.asExpr() = s.getKeySizeArg())
+  }
+
+  override predicate hasState(DataFlow::FlowState state) {
+    state = getMinAsymEcKeySize().toString()
+  }
+}
+
+/** A sink for an insufficient key size used in AES algorithms. */
+private class SymmetricSink extends InsufficientKeySizeSink {
+  SymmetricSink() {
+    exists(SymmetricInitMethodAccess ma, SymmetricKeyGenerator kg |
+      kg.getAlgoName() = "AES" and
+      DataFlow::localExprFlow(kg, ma.getQualifier()) and
+      this.asExpr() = ma.getKeySizeArg()
+    )
+  }
+
+  override predicate hasState(DataFlow::FlowState state) { state = getMinSymKeySize().toString() }
+}
+
+// ********************** SINKS HELPER CLASSES & PREDICATES **********************
+/** A call to a method that initializes a key generator. */
+abstract private class KeyGenInitMethodAccess extends MethodAccess {
```

Comment:
This class doesn't seem too useful. Its only value is to reuse the `getKeySizeArg` predicate, but the class itself isn't used anywhere. Besides, if in the future another 
class extending `KeyGenInitMethodAccess` was added, it may not have its keysize argument at position 0, so this would need refactoring anyway.

Not a strong opinion (your call), but I'd remove this and just add the predicate to `AsymmetricInitMethodAccess` and `SymmetricInitMethodAccess`.

Reply:
Updated in commit https://github.com/github/codeql/pull/10785/commits/8bc0a648637a91bd3c9a0e09516fb95eeba8a18c.